drivers/dose/dose.c

```diff
@@ -0,0 +1,566 @@
+            if (send_data_octet(ctx, *ptr)) {
+                goto collision;
+            }
+
+            /* Update CRC */
+            crc = crc16_update(crc, *ptr);
+
+            ptr++;
+        }
+    }
+
+    /* Send CRC */
+    network_uint16_t crc_nw = byteorder_htons(crc);
+    if (send_data_octet(ctx, crc_nw.u8[0])) {
+        goto collision;
+    }
+    if (send_data_octet(ctx, crc_nw.u8[1])) {
+        goto collision;
+    }
+
+    /* Send END octet */
+    if (send_octet(ctx, DOSE_OCTECT_END)) {
+        goto collision;
+    }
+
+    /* Get out of the SEND state */
+    state(ctx, DOSE_SIGNAL_END);
+
+    return pktlen;
+
+collision:
+    DEBUG("dose _send(): collision!\n");
+    if (--retries < 0) {
+        return 0;
+    }
+    goto send;
+}
+
+static int _get(netdev_t *dev, netopt_t opt, void *value, size_t max_len)
+{
+    dose_t *ctx = (dose_t *) dev;
+
+    switch (opt) {
+        case NETOPT_ADDRESS:
+            if (max_len < ETHERNET_ADDR_LEN) {
+                return -EINVAL;
+            }
+            memcpy(value, ctx->mac_addr, ETHERNET_ADDR_LEN);
+            return ETHERNET_ADDR_LEN;
+        case NETOPT_PROMISCUOUSMODE:
+            if (max_len < sizeof(netopt_enable_t)) {
+                return -EINVAL;
+            }
+            if (ctx->opts & DOSE_OPT_PROMISCUOUS) {
+                *((netopt_enable_t *)value) = NETOPT_ENABLE;
+            }
+            else {
+                *((netopt_enable_t *)value) = NETOPT_DISABLE;
+            }
+            return sizeof(netopt_enable_t);
+        default:
+            return netdev_eth_get(dev, opt, value, max_len);
+    }
+
+    return 0;
+}
+
+static int _set(netdev_t *dev, netopt_t opt, const void *value, size_t len)
+{
+    dose_t *ctx = (dose_t *) dev;
+
+    switch (opt) {
+        case NETOPT_PROMISCUOUSMODE:
+            if (len < sizeof(netopt_enable_t)) {
+                return -EINVAL;
+            }
+            if (((const bool *)value)[0]) {
+                SETBIT(ctx->opts, DOSE_OPT_PROMISCUOUS);
+            }
+            else {
+                CLRBIT(ctx->opts, DOSE_OPT_PROMISCUOUS);
+            }
+            return sizeof(netopt_enable_t);
+        default:
+            return netdev_eth_set(dev, opt, value, len);
+    }
+
+    return 0;
+}
+
+static int _init(netdev_t *dev)
+{
+    dose_t *ctx = (dose_t *) dev;
+    int irq_state;
+
+    /* Set state machine to defaults */
+    irq_state = irq_disable();
+    ctx->opts = 0;
+    ctx->recv_buf_ptr = 0;
+    ctx->flags = 0;
+    ctx->state = DOSE_STATE_UNDEF;
+    irq_restore(irq_state);
+
+    state(ctx, DOSE_SIGNAL_INIT);
+
+    return 0;
+}
+
+static const netdev_driver_t netdev_driver_dose = {
+    .send = _send,
+    .recv = _recv,
+    .init = _init,
+    .isr = _isr,
+    .get = _get,
+    .set = _set
+};
+
+void dose_setup(dose_t *ctx, const dose_params_t *params)
+{
+    ctx->netdev.driver = &netdev_driver_dose;
+
+    ctx->state = DOSE_STATE_UNDEF;
```

Comment:
This is set twice. `ctx->state` is also reset in the netdev driver's `init()` method.